### PR TITLE
Addresses HT-1196 update static IP

### PR DESCRIPTION
**Pull request by moseshll**
Add edit page and a whole lot of plumbing so that we have actual tests.

### CODE REVIEW

File: test/models/ht_user_test.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+require 'byebug'
+
+class HTUserTest < ActiveSupport::TestCase
```
Comment (aelkiss) [inline test/models/ht_user_test.rb]: I'd like to see some testing of the escaping/unescaping here.

Comment (moseshll) [reply test/models/ht_user_test.rb]: Test added

File: app/models/ht_user.rb [hunk 2]
```diff
@@ -1,5 +1,34 @@
 # frozen_string_literal: true
 
 class HTUser < ApplicationRecord
-  # self.table_name = "ht_users"
+  # Brain-dead IPv4 validation
+  validates :iprestrict, presence: true,
+                         format: { with: /\A\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\z/,
```
Comment (billdueber) [inline app/models/ht_user.rb]: Check out https://www.oreilly.com/library/view/regular-expressions-cookbook/9780596802837/ch07s16.html for a battle-tested ip regex

Comment (moseshll) [reply app/models/ht_user.rb]: Battle-tested regex swapped in, except ^ and $ replaced by \A and \z to keep Ruby happy.

File: app/views/shared/_flash_message.html.erb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<% unless flash.blank? %>
+  <% flash.each do |key, value| %>
+    <% alert_class = case key
```
Comment (billdueber) [inline app/views/shared/_flash_message.html.erb]: <priority very low> This seems like another place where we could use a translation file and use something like t("alert.css_class.#{key}", default: "alert-"#{key}"). In any case, it doesn't feel to me like it belongs here.

File: db/schema.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+# database schema. If you need to create the application database on another
+# system, you should be using db:schema:load, not running all the migrations
+# from scratch. The latter is a flawed and unsustainable approach (the more migrations
+# you'll amass, the slower it'll run and the greater likelihood for issues).
+#
+# It's strongly recommended that you check this file into your version control system.
+
+ActiveRecord::Schema.define(version: 0) do
+
+  create_table "ht_users", primary_key: "userid", id: :string, limit: 256, default: "", options: "ENGINE=MyISAM DEFAULT CHARSET=latin1", force: :cascade do |t|
```
Comment (billdueber) [inline db/schema.rb]: @aelkiss Are we not defaulting to innodb these days? Or not so much?

File: db/schema.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    t.string "displayname", limit: 128
+    t.string "email", limit: 128
+    t.string "activitycontact", limit: 128
+    t.string "approver", limit: 128
```
Comment (billdueber) [inline db/schema.rb]: If the userid can be 256, don't these need to be as well?

File: app/models/ht_user.rb (hunk 2 above)
Comment (billdueber) [reply app/models/ht_user.rb]: I'd use `\Z` (ignores trailing newline) and maybe add some `\s*` on the front and back so we don't break on leading/trailing spaces.

File: app/views/ht_users/index.html.erb [hunk 6]
```diff
@@ -1,32 +1,34 @@
...
+    <th>IP Restriction</th>
+    <th>MFA?</th>
+
+    <% @users.each do |u| %>
+      <tr>
+        <td><%= link_to u.userid, ht_user_path(u) %></td>
+        <td><%= u[:displayname] %></td>
+        <td><%= u[:email] %></td>
+        <td><%= u[:role] %></td>
+        <td><%= u[:expires]&.strftime("%Y-%m-%d") %></td>
```
[the remark this answers is not in the document]
Comment (moseshll) [reply app/views/ht_users/index.html.erb]: Artifact of development with placeholder DB contents, IIRC. The schema says NOT NULL so it's not likely to be an issue one way or the other.

File: app/models/ht_user.rb (hunk 2 above)
Comment (moseshll) [reply app/models/ht_user.rb]: Comments distilled into an edit of HT-1494.
Added a strip() to deal with whitespace for the current PR.

File: app/models/ht_user.rb [hunk 7]
```diff
@@ -1,5 +1,34 @@
...
+    Base64.encode64(userid)
+  end
+
+  # iprestrict is in the database as an escaped IPv4 regex e.g., ^127\.0\.0\.1$
+  # For the UI we strip out and restore the ^/./$ characters
+  def iprestrict
+    escaped = self[:iprestrict]
+    return escaped if escaped.nil?
+
+    escaped.gsub(/^\^/, '').gsub(/\$$/, '').gsub(/\\\./, '.')
```
[the remark this answers is not in the document]
Comment (moseshll) [reply app/models/ht_user.rb]: I think a full-blown regex was used so A&E could accommodate IP ranges. I'm thinking in particular of one I saw for the Copyright Office a while back. Aaron could probably dispel any myths I may be harboring.

File: app/controllers/ht_users_controller.rb [hunk 8]
```diff
@@ -1,16 +1,38 @@
...
   def index
     if params[:email]
       @users = HTUser.where('email LIKE ?', "%#{params[:email]}%").order(:userid)
+      flash.now[:alert] = "No results for '#{params[:email]}'" unless @users.count.positive?
     else
       @users = HTUser.all.order(:userid)
     end
   end
 
+  def update
```
[the remark this answers is not in the document]
Comment (moseshll) [reply app/controllers/ht_users_controller.rb]: Leaving that up to Jira.

File: app/controllers/ht_users_controller.rb [hunk 9]
```diff
@@ -1,16 +1,38 @@
 # frozen_string_literal: true
 
 class HTUsersController < ApplicationController
+  before_action :fetch_user, only: %i[show edit update]
+
   def index
     if params[:email]
       @users = HTUser.where('email LIKE ?', "%#{params[:email]}%").order(:userid)
+      flash.now[:alert] = "No results for '#{params[:email]}'" unless @users.count.positive?
```
[the remark this answers is not in the document]
Comment (moseshll) [reply app/controllers/ht_users_controller.rb]: Modified

File: app/models/ht_user.rb [hunk 10]
```diff
@@ -1,5 +1,34 @@
 # frozen_string_literal: true
 
 class HTUser < ApplicationRecord
-  # self.table_name = "ht_users"
+  # Brain-dead IPv4 validation
+  validates :iprestrict, presence: true,
+                         format: { with: /\A\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\z/,
+                                   message: 'requires a valid IPv4 address' }
+
+  HUMANIZED_ATTRIBUTES = {
```
[the remark this answers is not in the document]
Comment (moseshll) [reply app/models/ht_user.rb]: HT-1495 added.

File: app/views/shared/_flash_message.html.erb (hunk 3 above)
Comment (moseshll) [reply app/views/shared/_flash_message.html.erb]: twitter-bootstrap-rails gem might take care of this for us.

**Comment by aelkiss** [comment]
Regarding the regexes for IP ranges: The escaping of "." is required because "." matches any character by default in a regex. I agree the reason for the regexes is so that we can handle ranges or (potentially) multiple individual IP addresses for a person. That's definitely not the best way to deal with it, but changing the schema for the ht_users table is definitely out of scope here though as it would affect other components. I suspect we will eventually want to overhaul how we do user management entirely and at that time we can think about storing the IP ranges in a more robust way.

File: db/schema.rb (hunk 4 above)
Comment (aelkiss) [reply db/schema.rb]: I believe that we are. I think it's worth noting that db:setup would _never_ be run in production with this app - the table is created externally and this app just manages it.

File: db/schema.rb (hunk 5 above)
Comment (aelkiss) [reply db/schema.rb]: userid could be a shib persistent ID which is probably why it's longer. This schema is just copied from production, and changes to it are out of scope for this project.

File: app/views/ht_users/index.html.erb (hunk 6 above)
[the remark this answers is not in the document]
Comment (aelkiss) [reply app/views/ht_users/index.html.erb]: See comment below about regexes: I believe that we are at least currently, and changes to that are out of scope right now (but worth keeping in mind for if/when we redesign user management entirely)

File: app/models/ht_user.rb [hunk 11]
```diff
@@ -1,5 +1,34 @@
...
+    iprestrict: 'IP Restriction'
+  }.freeze
+
+  def self.human_attribute_name(attr, options = {})
+    HUMANIZED_ATTRIBUTES[attr.to_sym] || super
+  end
+
+  def to_param
+    Base64.encode64(userid)
```
[the remark this answers is not in the document]
Comment (moseshll) [reply app/models/ht_user.rb]: For some reason I can't make this work with /[^\/]+/ but it does work with /.+/ which seems like a nastier hack. Not going to try to modify this PR with it. It's HT-1497.

**Pull request merged**